docs(consul-api-gateway): add v0.4.0 upgrade instructions (#14101)
hashicorp/consul: website/content/docs/api-gateway/upgrades.mdx

This topic describes how to upgrade Consul API Gateway.

## Upgrade to v0.4.0

> Consul API Gateway v0.4.0 introduces support for [Gateway API v0.5.0](https://github.com/kubernetes-sigs/gateway-api/releases/tag/v0.5.0), the graduated v1beta1 `GatewayClass`, `Gateway` and `HTTPRoute` resources, and the new [`ReferenceGrant`](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.ReferenceGrant) resource, which is intended to replace the identical [`ReferencePolicy`](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.ReferencePolicy). This release includes backwards-compatibiltiy support for existing `ReferencePolicy` resources, but this will be removed in a future release, so we encourage users to migrate to `ReferenceGrant` after upgrading.

Author: This conflicts with some other just-merged changes now, will push up a commit manually breaking up these steps, then resolve this.

### Requirements

Ensure that the following requirements are met prior to upgrading:

- Consul API Gateway should be running version v0.3.0.

### Procedure

> 1. Complete the [standard upgrade](#standard-upgrade). After you've completed the upgrade, return to [post-upgrade configuration changes](#v0.4.0-post-upgrade-configuration-changes) to replace your `ReferencePolicy` resources with `ReferenceGrant` and upgrade your `GatewayClass`, `Gateway` and `HTTPRoute` resources from v1alpha2 to v1beta1.

Reviewer: ```suggestion
1. Complete the [standard upgrade](#standard-upgrade). 
1. After completing the upgrade, complete the [post-upgrade configuration changes](#v0.4.0-post-upgrade-configuration-changes). The post-upgrade procedure describes how to replace your `ReferencePolicy` resources with `ReferenceGrant` resources and how to upgrade your `GatewayClass`, `Gateway`, and `HTTPRoute` resources from v1alpha2 to v1beta1.
```
This is two steps.

> 1. Proceed with the [standard upgrade](#standard-upgrade), then after you've completed the upgrade, return to the [post-upgrade configuration changes instructions](#v0.4.0-post-upgrade-configuration-changes) to update your `GatewayClass`, `Gateway` and `HTTPRoute` resources from v1alpha2 to v1beta1, and replace your `ReferencePolicy` resources with `ReferenceGrant`.

Author: ```suggestion
1. Complete the [standard upgrade](#standard-upgrade). After you've completed the upgrade, return to [post-upgrade configuration changes](#v0.4.0-post-upgrade-configuration-changes) to replace your `ReferencePolicy` resources with `ReferenceGrant` and upgrade your `GatewayClass`, `Gateway` and `HTTPRoute` resources from v1alpha2 to v1beta1.
```

> 1. Proceed with the [standard upgrade](#standard-upgrade), then after you've completed the upgrade, return to the [post-upgrade configuration changes instructions](#v0.4.0-post-upgrade-configuration-changes) to update your `GatewayClass`, `Gateway` and `HTTPRoute` resources from v1alpha2 to v1beta1, and replace your `ReferencePolicy` resources with `ReferenceGrant`.

Author: Looks like this suggestion conflicts with a different one I just merged that similarly splits into two steps, but uses the terms `complete` instead of `perform` and `instructions` instead of `procedure` - could you resolve this and open a new suggestion against the current state?

<a name="v0.4.0-post-upgrade-configuration-changes"/>

> ### Post-Upgrade Configuration Changes

Reviewer: ```suggestion
### Post-upgrade configuration changes
```
The style guide was updated recently. Only page titles should use initial cap (title case) now.

#### Requirements

- Consul API Gateway should be running version v0.4.0.
- Consul Helm chart should be v0.47.0 or later.
- You should have the ability to run `kubectl` CLI commands.
- `kubectl` should be configured to point to the cluster containing the installation you are upgrading.

> - You should have the following permission rights on your Kubernetes cluster:

Reviewer 2: ```suggestion
- You should have the following permissions for your Kubernetes cluster:
```

  - `Gateway.read`
  - `ReferenceGrant.create` (Added in Consul Helm chart v0.47.0)
  - `ReferencePolicy.delete`

#### Procedure

1. Verify the current version of the `consul-api-gateway-controller` `Deployment`:

  ```shell-session
  $ kubectl get deployment --namespace consul consul-api-gateway-controller --output=jsonpath="{@.spec.template.spec.containers[?(@.name=='api-gateway-controller')].image}"
  ```

  You should receive a response similar to the following:

  ```log
  "hashicorp/consul-api-gateway:0.4.0"
  ```

  <a name="referencegrant"/>

1. Issue the following command to get all `ReferencePolicy` resources across all namespaces. If you have any active `ReferencePolicy` resources, you will receive output similar to the response below.

  ```shell-session
  $ kubectl get referencepolicy --all-namespaces
  Warning: ReferencePolicy has been renamed to ReferenceGrant. ReferencePolicy will be removed in v0.6.0 in favor of the identical ReferenceGrant resource.
  NAMESPACE   NAME
  default     example-reference-policy
  ```

>   If your output is empty, proceed to upgrading your `GatewayClass`, `Gateway` and `HTTPRoute` resources to [v1beta1](#v1beta1-gatewayclass-gateway-httproute).

Reviewer: ```suggestion
  If your output is empty, upgrade your `GatewayClass`, `Gateway` and `HTTPRoute` resources to v1beta1 as described in [step 7](#v1beta1-gatewayclass-gateway-httproute).
```
Just so readers can make sure they're in the right place if they get lost.

> 1. For each `ReferencePolicy`, edit the source YAML file, updating to `kind: ReferenceGrant` and optionally updating the `metadata.name` field or filename if they previously included the term "policy", as in the example below. Note that because the `kind` field is being updated, it is not possible to use `kubectl edit` to edit the remote state directly.

Reviewer: ```suggestion
1. For each `ReferencePolicy` in the YAML file, change the `kind` field to `ReferenceGrant`. You can optionally update the `metadata.name` field or filename if they include the term "policy". In the following example, the `kind` and `metadata.name` fields have been changed to the new resource. Note that updating the `kind` field prevents you from using the `kubectl edit` command to edit the remote state directly. 
```
Trying to improve the flow; better?

Author: I think specifying `source YAML` (as preserved in the suggestion below for v1beta1 upgrades) is relevant to contrast with editing the live resource on the Kubernetes server - will add that word back and merge this.

  spec:
    from:
    - group: gateway.networking.k8s.io
      kind: HTTPRoute
      namespace: example-namesapce
    to:
    - group: ""
      kind: Service
      name: web-backend
  ```

  </CodeBlockConfig>

1. For each file, apply the updated YAML to your cluster to create a new `ReferenceGrant` resource.

  ```shell-session
  $ kubectl apply -f <file>
  ```

1. Check to confirm that each new `ReferenceGrant` was created successfully.

  ```shell-session
  $ kubectl get referencegrant <name> --namespace <namespace>
  NAME
  example-reference-grant
  ```

1. Finally, delete each corresponding old `ReferencePolicy` resource. Because replacement `ReferenceGrant` resources have already been created, there should be no interruption in the availability of any referenced `Service` or `Secret`.

  ```shell-session
  $ kubectl delete referencepolicy <name> --namespace <namespace>
  Warning: ReferencePolicy has been renamed to ReferenceGrant. ReferencePolicy will be removed in v0.6.0 in favor of the identical ReferenceGrant resource.
  referencepolicy.gateway.networking.k8s.io "example-reference-policy" deleted
  ```

  <a name="v1beta1-gatewayclass-gateway-httproute"/>

> 1. For each `GatewayClass`, `Gateway`, and `HTTPRoute`, edit the source YAML file, updating to `apiVersion: gateway.networking.k8s.io/v1beta1`. Note that because the `apiVersion` field is being updated, it is not possible to use `kubectl edit` to edit the remote state directly.

Reviewer: ```suggestion
1. For each `GatewayClass`, `Gateway`, and `HTTPRoute` object in the source YAML, update the `apiVersion` field to `gateway.networking.k8s.io/v1beta1`. Note that updating the `apiVersion` field prevents you from using the `kubectl edit` command to edit the remote state directly.
```
Now that I'm rereading that note about the `kubectl edit` command, I'm not sure that this is what we mean, either. Is it that certain fields (`apiVersion` and `kind`) are not editable or accessible from the command? The way it's worded, it seems as if the simple act of changing those fields is the reason why you can't use the command.

Author: > Is it that certain fields (apiVersion and kind) are not editable or accessible from the command?

Yes, that's correct, it will return an error and fail to update the remote resource.

Author: Removed the word `object` to align with suggested language for the ReferencePolicy update above.

Consul API Gateway v0.4.0 is backward-compatible with existing `ReferencePolicy` resources, but we will remove support for `ReferencePolicy` resources in a future release. We recommend that you migrate to `ReferenceGrant` after upgrading.

### Requirements

Ensure that the following requirements are met prior to upgrading:

- Consul API Gateway should be running version v0.3.0.

### Procedure

1. Complete the [standard upgrade](#standard-upgrade).
1. After completing the upgrade, complete the [post-upgrade configuration changes](#v0.4.0-post-upgrade-configuration-changes). The post-upgrade procedure describes how to replace your `ReferencePolicy` resources with `ReferenceGrant` resources and how to upgrade your `GatewayClass`, `Gateway`, and `HTTPRoute` resources from v1alpha2 to v1beta1.

<a name="v0.4.0-post-upgrade-configuration-changes"/>

### Post-upgrade configuration changes

> 

Reviewer: ```suggestion
Complete the following steps after performing standard upgrade procedure. 
```
Just so we don't have headings stacked on top of headings per the style guide.

#### Requirements

- Consul API Gateway should be running version v0.4.0.
- Consul Helm chart should be v0.47.0 or later.
- You should have the ability to run `kubectl` CLI commands.
- `kubectl` should be configured to point to the cluster containing the installation you are upgrading.
- You should have the following permissions for your Kubernetes cluster:
  - `Gateway.read`
  - `ReferenceGrant.create` (Added in Consul Helm chart v0.47.0)
  - `ReferencePolicy.delete`

#### Procedure

1. Verify the current version of the `consul-api-gateway-controller` `Deployment`:

  ```shell-session
  $ kubectl get deployment --namespace consul consul-api-gateway-controller --output=jsonpath="{@.spec.template.spec.containers[?(@.name=='api-gateway-controller')].image}"
  ```

  You should receive a response similar to the following:

  ```log
  "hashicorp/consul-api-gateway:0.4.0"
  ```

  <a name="referencegrant"/>

> 1. Issue the following command to get all `ReferencePolicy` resources across all namespaces. If you have any active `ReferencePolicy` resources, you will receive output similar to the response below.
> 
>   ```shell-session
>   $ kubectl get referencepolicy --all-namespaces
>   Warning: ReferencePolicy has been renamed to ReferenceGrant. ReferencePolicy will be removed in v0.6.0 in favor of the identical ReferenceGrant resource.
>   NAMESPACE   NAME
>   default     example-reference-policy
>   ```

Reviewer 3: ```suggestion
1. Issue the following command to get all `ReferencePolicy` resources across all namespaces.

  ```shell-session
  $ kubectl get referencepolicy --all-namespaces
  ```
If you have any active `ReferencePolicy` resources, you will receive output similar to the response below.
```log
  Warning: ReferencePolicy has been renamed to ReferenceGrant. ReferencePolicy will be removed in v0.6.0 in favor of the identical ReferenceGrant resource.
  NAMESPACE   NAME
  default     example-reference-policy
  ```
```
